### ISSUE TITLE
Predictable image formation for HDR color spaces

### ISSUE THREAD

**Issue by donmccurdy** [gpuweb/gpuweb#4919]
Within the recent updates for HDR support in WebGPU, applications are required to target either Extended sRGB or Extended Display P3 color spaces, within which:

> ... colors will not be restricted to the standard dynamic range, but rather will be restricted to the full dynamic range of the display device.

Importantly, web applications do not (and presumably cannot, privacy reasons?) know the full dynamic range of the display device. I've read the attestations in https://github.com/gpuweb/gpuweb/issues/4239 that unclipped sRGB and/or Display P3 will enable major HDR use cases, but I'm concerned that there's an important missing piece here, both conceptually and in practice.

In a typical 3D rendering pipeline we model light propagation within an RGB stimulus encoding, and then apply an image formation transform (of which "tone mapping" may be a part) to create a picture. The second step is essential, not only because the dynamic range of the display is limited, but more fundamentally because pictures exist relative to a medium. Since HDR support landed in Chrome 129, I've seen a half dozen HDR demos in various WebGPU-based engines, and all have disabled their image formation steps in the renders. While some misunderstanding of HDR and image formation may be a part of that, the currently-shipped spec also does not define the medium against which we are forming a picture. While sRGB [0,1] may behave predictably, we have no insight into what the OS and/or display are doing with the encoded values >1, and this varies dramatically, [even for a single display](https://www.instagram.com/p/C_Byx6jRfuX/) (on a recent macOS display, try adjusting brightness here).

I'm not certain what API surface is needed to enable predictable image formation targeting HDR-capable displays; this may lead into discussion of OS-level behaviors. But I am concerned that WebGPU-based applications may start to ship with broken assumptions about image formation and HDR displays, which would be harder to fix later on. Perhaps Rec. 2100 HLG, which does define nominal peak luminance, might behave more predictably than Extended sRGB, though I currently have no way to test that theory. I see `rec2100-display-linear` also mentioned in the explainer, though I'm uncertain how/if that differs from Linear Rec. 2020, which would present similar problems.

_Disclaimer: I'm a contributor to [three.js](https://threejs.org/)._

**Comment by beaufortfrancois** [comment]
FYI @ccameron-chromium

**Comment by ccameron-chromium** [comment]
> Importantly, web applications do not (and presumably cannot, privacy reasons?) know the full dynamic range of the display device.

Yes, exactly. To the extent that this will ever be exposed, it will not be reliably available.

The goal here is to provide more `mode` options to allow custom tone mapping, so that an application can say, "I'm going to write values in the [0,5] interval, browser, you do this specific tone mapping algorithm to get it into the available HDR headroom".

Exactly what values of `mode` to provide is a bit harder a problem, and so we're taking the approach of providing this API with the two values that were uncontroversial ("clamp to standard" and "clamp to whatever the display supports") to make progress on the plumbing.

**Comment by magcius** [comment]
Clamping the available range will result in gamut clipping, causing the ["Notorious Six" effect](https://alextardif.com/Tonemapping.html). Image formation (aka tonemapping) is inherently a lossy procedure, which means the choice of algorithm should be in the hands of the application so it can use the range effectively.

Metadata approaches about the content range has been tried before, without success; SMPTE ST.2086 is widely considered a failure.

**Comment by donmccurdy** [comment]
Thanks @ccameron-chromium and @magcius!

I think we're using the same term, "tone mapping", but referring to at least two different things.  I'll try to disambiguate with "image formation" vs "output device transform":

<img width="1406" alt="image_pipeline" src="https://github.com/user-attachments/assets/4e74dd43-da36-47e2-8364-203cd99e9afc">

The application must do the first steps, perhaps implementing image formation with common tone mappers like ACES Filmic, Reinhard, or AgX, using some well-defined HDR reference display standard as the target medium for the image. Then the browser / OS must adapt that image to the user's actual display with an output device transform.

Configurability aside, are we in agreement so far?

**Comment by magcius** [comment]
Possibly, but I don't see value in image formation being used with an arbitrary reference display. With more modern tonemappers like AgX or glTF Commercial that provide a better path to white, the amount of gamut compression should likely be influenced by the MaxFFTML of the final output display to limit clipping artifacts. To have a smooth, non-clipped roll-off, you need to know when your curve should start and end.

**Comment by donmccurdy** [comment]
It's not my decision to make, but I have a very hard time imagining that browser vendors will expose information about the user's personal display. Without that, we still have to target something. Despite variation in displays, we've had a [defined target for SDR images](https://www.w3.org/Graphics/Color/sRGB.html#:~:text=The%20sRGB%20reference%20viewing%20environment,of%20the%20maximum%20white%20luminance.) for some time. The fundamentals do not change because the displays are now brighter — a target is still useful, just as HDR video can be produced in a small handful of formats with wide support.

Note that the Khronos PBR Neutral (“Commerce”) tone mapper is not applicable to HDR or wide-gamut displays, but I believe AgX can be adapted to either.

**Comment by samdenty** [comment]
relevant issue: https://github.com/w3c/csswg-drafts/issues/10110

**Comment by samdenty** [comment]
also https://github.com/w3c/ColorWeb-CG/issues/113

**Comment by donmccurdy** [comment]
I've been reviewing requirements to enable HDR support for [three.js](https://threejs.org/), and unfortunately this issue remains a major obstacle. I'm not confident that three.js (or other similar projects) can go past highly 'experimental' support for HDR, without an encoding that is proportionate (0–100%) within some limits.

I would really prefer to see Rec. 2100 HLG supported. For example:

```javascript
context.configure({
    device: device,
    usage: GPUTextureUsage.RENDER_ATTACHMENT,
    alphaMode: "opaque",
    colorSpace: "rec2100-hlg",
    format: "rgba16float", // "rgba16float" | "rgb10a2unorm"
});
```

Is this still a direction that WebGPU authors are considering? Thank you! 🙏🏻

**Comment by kainino0x** [comment]
From our side @ccameron-chromium would need to weigh in since he better understands what HLG is.

We pick up our color space list [from the HTML spec](https://html.spec.whatwg.org/multipage/canvas.html#predefinedcolorspace) which inherits definitions from [the CSS Color Module Level 4 spec](https://www.w3.org/TR/css-color-4/#predefined). It only has `rec2020` right now. But CSS color spaces are extended, and Rec.2100 has the same primaries as Rec.2020, so maybe we get what is needed without adding `rec2100`? I am unsure know how the transfer function difference would interact with us....

**Comment by ccameron-chromium** [comment]
> I think we're using the same term, "tone mapping", but referring to at least two different things. I'll try to disambiguate with "image formation" vs "output device transform"

Yes, this is a very subtle and important issue. The application is responsible for creating a display-referred image (part of which may involve an operation to convert from a scene-referred-ish light representation, which can in some contents be called "tone mapping").

The browser is responsible for a "display referred to display referred" tone mapping. This tone mapping will take the form of "given an image with HDR headroom X, tone map it to HDR headroom Y" (where X is specified by the content, and Y is a combination of display capabilities and things like `dynamic-range-limit`).

We plan to expose an API whereby the application can specify a custom curve to apply, along with parameters for how the curve is to be applied, for any value of Y.

In terms of HLG and PQ spaces, they can be useful for using fixed-point instead of floating-point. But we want the color space and the tone mapping to be completely independent (so you can specify the same content via HLG, PQ, linear, sRGB, or Display P3).

**Comment by donmccurdy** [comment]
Thanks @ccameron-chromium and @kainino0x!

About keeping WebGPU's color space and tone mapping concepts independent, I fully agree.

If an application specifies a custom curve, would that allow a mapping from:

- (a) [0, X] → [0, Y] (entire image) (EDIT: preferred)
- (b) [1, X] → [1, Y] (headroom only)

In other words, does this strictly require that the domain [0, 1] be unaffected by any change in Y?

**Comment by ccameron-chromium** [comment]
> If an application specifies a custom curve, would that allow a mapping from:
> 
> * (a) [0, X] → [0, Y] (entire image)
> * (b) [1, X] → [1, Y] (headroom only)
> 
> In other words, does this strictly require that the domain [0, 1] be unaffected by any change in Y?

The behavior will be fully configurable. So you *can* leave [0, 1] unaffected (at the expense of doing some violence to values >1), or you can leave, say, [0, 0.5] unaffected, or do a linear scaling in [0, 1] followed by some toe curve.

The current `"extended"` mode is "just clamp everything that doesn't fit", because that was the only useful behavior we were ready to document.

**Comment by stephtr** [comment]
Sorry for chiming in with my naivety..

For still images, gain maps are an effective solution for HDR under varying display conditions. _(An SDR base image is stored together with a per-pixel scaling factor to reconstruct the HDR image at a fixed maximum dynamic range. When displayed, the scaling factors are dynamically adjusted to the device's HDR headroom.)_

There are some use cases which would benefit from applying a similar approach to real-time rendering in WebGPU. But I guess the complexity and likely small user base is a dealbreaker here?

**Comment by ccameron-chromium** [comment]
> For still images, gain maps are an effective solution for HDR under varying display conditions. _(An SDR base image is stored together with a per-pixel scaling factor to reconstruct the HDR image at a fixed maximum dynamic range. When displayed, the scaling factors are dynamically adjusted to the device's HDR headroom.)_
> 
> There are some use cases which would benefit from applying a similar approach to real-time rendering in WebGPU. But I guess the complexity and likely small user base is a dealbreaker here?

Yes, gain maps are an excellent technology. Indeed their complexity is what made them not be the go-to option here. The likely way to handle them would be to expose a separate rendertarget for the "alternate image" (in the parlance ISO 21496-1), rather than having a gain map image.

For the moment, though, we're going with a global tone mapping that is common to images, video, and canvas.

**Comment by donmccurdy** [comment]
The relative encoding is the really compelling feature of HLG. A realtime renderer can then form the display-referred image with respect to relative limits, \[0, X]. If the result is then scaled proportionally onto \[0, Y], all is well. Where we'll run into trouble is if, after forming a picture within \[0, X], the \[1, X] interval (or any other interval) is mapped differently from the rest of the picture. A short recording, [Luminance... We need to talk](https://vimeo.com/651149199) by Daniele Siragusano brings up some of the challenges (not specific to HDR encodings), but briefly — gradients and ratios are critical parts of the picture's encoding, including their arrangement in the viewport and in time.

Custom tone mapping would allow us to display HLG-like relative encodings proportionally, if I understand you correctly, with something like this configuration:<sup>1</sup>

- colorSpace = "rec2100-hlg"
- toneMapping.mode = "extended"
- toneMapping.??? = \<implementation linearly scaling \[0, X] → \[0, Y]>

This might be OK, just a bit custom-feeling, for the purpose of showing HLG-like content as authored. WebGPU aside, is HLG video handled differently? Have I departed from the spec's “happy path” here, and if so, is there a reason for that?

Alternatively applications could defer their own tone mapping until this new, custom tone mapping function you describe. This would require at minimum a 16-bit drawing buffer (with performance implications), and arbitrary [RGB-to-RGB linear algebra](https://iolite-engine.com/blog_posts/minimal_agx_implementation). I'm not sure if you intend to allow custom tone mapping that much flexibility?

***

<small><sup>1</sup> Ignoring the Rec. 2020 primaries used in HLG, for now, Rec. 709 or P3 would also serve.</small>

**Comment by kainino0x** [comment]
Closing as this needs to get standardized at the broader [Color on the Web CG](https://github.com/w3c-cg/ColorWeb-CG).
